### PR TITLE
Simplify the logic to update the extensions goal state

### PR DESCRIPTION
**Pull request by narrieta**

### CODE REVIEW

File: azurelinuxagent/common/protocol/wire.py [hunk 1]
```diff
@@ -577,7 +577,7 @@ def __init__(self, endpoint):
         self._endpoint = endpoint
         self._goal_state = None
         self._extensions_goal_state = None  # The goal state to use for extensions; can be an ExtensionsGoalStateFromVmSettings or ExtensionsGoalStateFromExtensionsConfig
-        self._vm_settings_goal_state = None  # Cached value of the most recent ExtensionsGoalStateFromVmSettings
```
Comment (narrieta) [inline azurelinuxagent/common/protocol/wire.py:580]: I simplified the logic by reducing the state we keep for this operation, reducing the number of code branches, simplifying the error handling, and updating the extensions goal state in 1 single place. I also added a few comments.

I renamed this variable and reduced the scope of its usage. Now it is only used for caching purposed (and to save the cached value for debugging purposes). It does not play a role in setting the extensions goal state anymore.

File: azurelinuxagent/common/protocol/wire.py [hunk 2]
```diff
@@ -806,42 +813,35 @@ def update_goal_state(self, force_update=False):
...
+                        self._vm_settings_error_reporter.report_error(format_exception(error))
+                self._vm_settings_error_reporter.report_summary()
+
+            #
+            # If we fetched a valid vmSettings, use that for the extensions goal state. Otherwise use ExtensionsConfig
+            #
+            if vm_settings_goal_state_is_valid:
+                self._extensions_goal_state = vm_settings_goal_state
             else:
-                if not self._host_plugin_supports_vm_settings and self._host_plugin_supports_vm_settings_next_check > datetime.now():
```
Comment (narrieta) [inline azurelinuxagent/common/protocol/wire.py:814]: self._host_plugin_supports_vm_settings and self._host_plugin_supports_vm_settings_next_check were moved to _fetch_vm_settings_goal_state()

File: azurelinuxagent/common/protocol/wire.py [hunk 3]
```diff
@@ -806,42 +813,35 @@ def update_goal_state(self, force_update=False):
...
-                    self._extensions_goal_state = self._goal_state.extensions_config
-                else:
-                    try:
-                        vm_settings_goal_state = self._fetch_vm_settings_goal_state(force_update=force_update)
-
-                        self._host_plugin_supports_vm_settings = True
-
-                        if self._vm_settings_goal_state is None or self._vm_settings_goal_state.etag != vm_settings_goal_state.etag:
-                            self._vm_settings_goal_state = vm_settings_goal_state
-                            self._extensions_goal_state = vm_settings_goal_state
```
Comment (narrieta) [inline azurelinuxagent/common/protocol/wire.py:825]: Now self._extensions_goal_state is updated only once, after all the logic to decide which goal state to use has been executed

File: azurelinuxagent/common/protocol/wire.py [hunk 4]
```diff
@@ -854,10 +854,25 @@ def update_goal_state(self, force_update=False):
...
+        Raises ProtocolError if the request fails for any reason (e.g. not supported, time out, server error)
         """
-        etag = None if force_update or self._vm_settings_goal_state is None else self._vm_settings_goal_state.etag
+        def raise_not_supported(reset_state=False):
+            if reset_state:
+                self._host_plugin_supports_vm_settings = False
+                self._host_plugin_supports_vm_settings_next_check = datetime.now() + timedelta(hours=6)  # check again in 6 hours
+                # "Not supported" is not considered an error, so don't use self._vm_settings_error_reporter to report it
+                logger.info("vmSettings is not supported")
+                add_event(op=WALAEventOperation.HostPlugin, message="vmSettings is not supported", is_success=True)
```
Comment (larohra) [inline azurelinuxagent/common/protocol/wire.py:868]: Shouldn't the op be VMSettings?

```suggestion
                add_event(op=WALAEventOperation.VmSettings, message="vmSettings is not supported", is_success=False)
```

**Review by larohra — APPROVED** [review]
1 minor comment else LGTM

**Review by kevinclark19a — APPROVED** [review]

File: azurelinuxagent/common/protocol/wire.py (hunk 4 above)
Comment (narrieta) [reply azurelinuxagent/common/protocol/wire.py:868]: I used VmSettings for all the telemetry we need to monitor (errors, number of requests) and HostGAPlugin for informational purposes (HostGAPlugin version, vmSettings supported)

**Pull request merged**